flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java

```diff
@@ -0,0 +1,801 @@
+			LOG.warn("TaskManager resource registration failed for {}", resourceID);
+
+			// tell the JobManager about the failure
+			String eStr = ExceptionUtils.stringifyException(e);
+			sender().tell(decorateMessage(
+				new RegisterResourceFailed(taskManager, resourceID, eStr)), self());
+		}
+	}
+
+	/**
+	 * Releases the given resource. Note that this does not automatically shrink
+	 * the designated worker pool size.
+	 *
+	 * @param resourceId The TaskManager's resource id.
+	 */
+	private void removeRegisteredResource(ResourceID resourceId) {
+
+		WorkerType worker = registeredWorkers.remove(resourceId);
+		if (worker != null) {
+			releaseRegisteredWorker(worker);
+		} else {
+			LOG.warn("Resource {} could not be released", resourceId);
+		}
+	}
+
+
+	// ------------------------------------------------------------------------
+	//  Registration and consolidation with JobManager Leader
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Called as soon as we discover (via leader election) that a JobManager lost leadership
+	 * or a different one gained leadership.
+	 *
+	 * @param leaderAddress The address (Akka URL) of the new leader. Null if there is currently no leader.
+	 * @param leaderSessionID The unique session ID marking the leadership session.
+	 */
+	protected void newJobManagerLeaderAvailable(String leaderAddress, UUID leaderSessionID) {
+		log.debug("Received new leading JobManager {}. Connecting.", leaderAddress);
+
+		// disconnect from the current leader (no-op if no leader yet)
+		jobManagerLostLeadership();
+
+		// a null leader address means that only a leader disconnect
+		// happened, without a new leader yet
+		if (leaderAddress != null) {
+			// the leaderSessionID implicitly filters out success and failure messages
+			// that come after leadership changed again
+			this.leaderSessionID = leaderSessionID;
+			triggerConnectingToJobManager(leaderAddress);
+		}
+	}
+
+	/**
+	 * Causes the resource manager to announce itself at the new leader JobManager and
+	 * obtains its connection information and currently known TaskManagers.
+	 *
+	 * @param leaderAddress The akka actor URL of the new leader JobManager.
+	 */
+	private void triggerConnectingToJobManager(String leaderAddress) {
+		LOG.info("Trying to associate with JobManager leader " + leaderAddress);
+
+		final Object registerMessage = decorateMessage(new RegisterResourceManager(self()));
+		final Object retryMessage = decorateMessage(new TriggerRegistrationAtJobManager(leaderAddress));
+
+		// send the registration message to the JobManager
+		ActorSelection jobManagerSel = context().actorSelection(leaderAddress);
+		Future<Object> future = Patterns.ask(jobManagerSel, registerMessage, new Timeout(messageTimeout));
+
+		future.onComplete(new OnComplete<Object>() {
+
+			@Override
+			public void onComplete(Throwable failure, Object msg) {
+				if (msg != null) {
```

Comment:
I think it would be better to first check whether `failure != null` which indicates a failure case. Furthermore, I would log this `Throwable` with the failure message.